applications/nrf_desktop/src/modules/dfu_mcumgr.c

```diff
@@ -57,42 +58,45 @@ static void dfu_timeout_handler(struct k_work *work)
 	}
 }
 
-static enum mgmt_cb_return mcumgr_img_mgmt_cb(uint32_t event,
-					      enum mgmt_cb_return prev_status,
-					      int32_t *rc, uint16_t *group,
-					      bool *abort_more, void *data,
-					      size_t data_size)
+static enum mgmt_cb_return smp_cmd_recv(uint32_t event, enum mgmt_cb_return prev_status,
+					int32_t *rc, uint16_t *group, bool *abort_more,
+					void *data, size_t data_size)
 {
-	LOG_DBG("MCUmgr Image Management Event with the %d ID",
-		u32_count_trailing_zeros(MGMT_EVT_GET_ID(event)));
+	LOG_INF("MCUmgr SMP Command Recv Event");
 
-	k_work_reschedule(&dfu_timeout, DFU_TIMEOUT);
-	if (IS_ENABLED(CONFIG_DESKTOP_DFU_LOCK) && dfu_lock_claim(&mcumgr_owner)) {
-		(void)k_work_cancel_delayable(&dfu_timeout);
-		*rc = MGMT_ERR_EACCESSDENIED;
-		return MGMT_CB_ERROR_RC;
-	}
+	const struct mgmt_evt_op_cmd_arg *cmd_recv;
 
 	if (IS_ENABLED(CONFIG_MCUMGR_TRANSPORT_BT) &&
 	    atomic_cas(&mcumgr_event_active, false, true)) {
 		APP_EVENT_SUBMIT(new_ble_smp_transfer_event());
 	}
 
-	return MGMT_CB_OK;
-}
+	if (!IS_ENABLED(CONFIG_DESKTOP_DFU_LOCK)) {
+		return MGMT_CB_OK;
+	}
 
-static struct mgmt_callback img_mgmt_callback = {
-	.callback = mcumgr_img_mgmt_cb,
-	.event_id = MGMT_EVT_OP_IMG_MGMT_ALL,
-};
+	if (event != MGMT_EVT_OP_CMD_RECV) {
+		LOG_ERR("Spurious event in recv cb: %" PRIu32, event);
+		*rc = MGMT_ERR_EUNKNOWN;
+		return MGMT_CB_ERROR_RC;
+	}
 
-static enum mgmt_cb_return mcumgr_os_mgmt_reset_cb(uint32_t event,
-						   enum mgmt_cb_return prev_status,
-						   int32_t *rc, uint16_t *group,
-						   bool *abort_more, void *data,
-						   size_t data_size)
-{
-	LOG_INF("MCUmgr OS Management Reset Event");
+	if (data_size != sizeof(*cmd_recv)) {
+		LOG_ERR("Invalid data size in recv cb: %zu (expected: %zu)",
+			data_size, sizeof(*cmd_recv));
+		*rc = MGMT_ERR_EUNKNOWN;
+		return MGMT_CB_ERROR_RC;
+	}
+
+	cmd_recv = data;
```

Comment:
This may be done before submitting the `ble_smp_transfer_event` (on error the seems to be no need to decrease latency)

Reply:
Consider moving the validation logic to a dedicated (2 code blocks + 1 suggested by Marek). It can be function returning the bool value if you return the same error data in each failing step of validation blocks

Reply:
> You can also check if `cmd_recv.err` is `MGMT_ERR_EOK` to make the validation complete

It is not used in the recv callback

```
/**
 * Arguments for #MGMT_EVT_OP_CMD_RECV, #MGMT_EVT_OP_CMD_STATUS and #MGMT_EVT_OP_CMD_DONE
 */
struct mgmt_evt_op_cmd_arg {
	/** #mcumgr_group_t */
	uint16_t group;

	/** Message ID within group */
	uint8_t id;

	union {
		/** #mcumgr_err_t, used in #MGMT_EVT_OP_CMD_DONE */
		int err;

		/** #img_mgmt_id_upload_t, used in #MGMT_EVT_OP_CMD_STATUS */
		int status;
	};
};
```

Reply:
> Consider moving the validation logic to a dedicated (2 code blocks + 1 suggested by Marek). It can be function returning the bool value if you return the same error data in each failing step of validation blocks

In case of 3 block I would move it, however as it will stay at 2 code blocks (reason in comment before) imho it is cleaner to keep it in-lined in this function.

Reply:
> > You can also check if `cmd_recv.err` is `MGMT_ERR_EOK` to make the validation complete
> 
> It is not used in the recv callback
> 
> ```
> /**
>  * Arguments for #MGMT_EVT_OP_CMD_RECV, #MGMT_EVT_OP_CMD_STATUS and #MGMT_EVT_OP_CMD_DONE
>  */
> struct mgmt_evt_op_cmd_arg {
> 	/** #mcumgr_group_t */
> 	uint16_t group;
> 
> 	/** Message ID within group */
> 	uint8_t id;
> 
> 	union {
> 		/** #mcumgr_err_t, used in #MGMT_EVT_OP_CMD_DONE */
> 		int err;
> 
> 		/** #img_mgmt_id_upload_t, used in #MGMT_EVT_OP_CMD_STATUS */
> 		int status;
> 	};
> };
> ```

Nice catch. The code actually sets the value, but we should not rely on it since documentation does not mention it.

Reply:
ok, resolving